doc/specs/stdlib_hash_maps.md

```diff
@@ -0,0 +1,3088 @@
+`key_type`, and `other_type`. The `key_type` is intended to be used
+for the search keys of hash tables.  The `other_type` is intended to
+store additional data associated with a key. Both types are
+opaque. Their current representations are as follows
+
+```fortran
+    type :: key_type
+        private
+        integer(int8), allocatable :: value(:)
+    end type key_type
+
+    type :: other_type
+        private
+        integer(int8), allocatable :: value(:)
+    end type other_type
+```
+
+The  module also defines seven procedures for those types: `copy_key`,
+`copy_other`, `free_key`, `free_other`, `get`, `key_test`, and `set`
+for use by the hash maps to manipulate or inquire of components of
+those types.
+
+### Table of `stdlib_32_bit_key_data_wrapper` procedures
+
+The  `stdlib_32_bit_key_data_wrapper` module provides procedures in
+several categories: procedures to manipulate data of the `key_type`;
+procedures to manipulate data of the `other_type`, and 32 bit hash
+functions for keys. The procedures in each category are listed below.
+
+Procedures to manipulate `key_type` data:
+
+* `copy_key( key_in, key_out )` - Copies the contents of the key,
+  key_in, to the key, key_out.
+
+* `get( key, value )` - extracts the content of key into value.
+
+* `free_key( key )` - frees the memory in key.
+
+* `set( key, value )` - sets the content of key to value.
+
+* `key_test( key1, key2 )` - compares two keys for equality.
+
+Procedures to manipulate `other_type` data:
+
+* `copy_other( other_in, other_out )` - Copies the contents of the
+  other data in, other_in, to the other data, other_out.
+
+* `get( other, value )` - extracts the content of other into value.
+
+* `set( other, value )` - sets to content of other to value.
+
+* `free_other( other )` - frees the memory in other.
+
+Procedures to hash keys to 32 bit integers:
+
+* `fnv_1_hasher( key )` - hashes a key using the FNV-1 algorithm.
+
+* `fnv_1a_hasher( key )` - hashes a key using the FNV-1a algorithm.
+
+* `seeded_nmhash32_hasher( key )` - hashes a key using the nmhash32
+  algorithm.
+
+* `seeded_nmhash32x_hasher( key )` - hashes a key using the nmhash32x
+  algorithm.
+
+* `seeded_water_hasher( key )` - hashes a key using the waterhash
+  algorithm.
+
+### Specifications of the `stdlib_32_bit_key_data_wrapper` procedures
+
+#### `copy_key` - Returns a copy of the key
+
+##### Status
+
+Experimental
+
+##### Description
+
+Returns a copy of an input of type `key_type`
+
+##### Syntax
+
+`call [[stdlib_32_bit_key_data_wrapper:copy_key]]( key_in, key_out )`
+
+##### Class
+
+Subroutine.
+
+##### Arguments
+
+`key_in`: shall be a scalar expression of type `key_type`. It
+is an `intent(in)` argument.
+
+`key_out`: shall be a scalar variable of type `key_type`. It
+is an `intent(out)` argument.
+
+##### Example
+
+```fortran
+    program demo_copy_key
+      use stdlib_32_bit_key_data_wrapper, only: &
+          copy_key, key_test, key_type
```

Comment:
It looks like `key_test` is a test for equality? In that case consider if it would be better named `equal_keys`.

Reply:
> > The current use of `inmap` is about a factor if two more efficient than using `key` directly for modifications of the `other` data associated with the `key`. For initially entering an `entry`, testing for existence of the `key` in the `map` , and deleting an `entry` from the `map` there is not much difference in the efficiency between using `inmap` and using `key` directly.
> 
> If querying the key twice is a performance issue, could you return a `pointer` to other rather than its index `inmap`?

I would not return a pointer to internal data. It can cause a memory leak or other problems, and would require making more of the internal `targets` or `pointers`, making the structure more fragile. It is a performance issue for a lesser used aspect of hash maps, and I would rather make a localized performance degradation than use pointers.

Reply:
I don't know why this comment appeared here and not on the tail of the comments.